### DISCUSSION TITLE
Why isn't RandomBooleanSequenceGenerator the default behaviour?

### DISCUSSION THREAD

**Discussion by airzym** [AutoFixture/AutoFixture#1283] [Q&A] [answered]
Sorry if this topic has already been covered, but I couldn't find the answer in previous discussions/issues.

Can anyone shed any light for me on why the `RandomBooleanSequenceGenerator` isn't the default behaviour for generating boolean values? 

For me the beauty of AutoFixture is that it allows me to generate any and all permutations of parameters, however when boolean values are always generated with the same value each time, it reduces the value of using AutoFixture.

I had the following bug today, which the default behaviour "hid" from me, and without understanding how AutoFixture generated booleans by default, it may not be immiditaly obvious to anyway why the test always passed.

```C#
public class Foo
{
    public Foo(bool one, bool two)
    {
        One = one;
        Two = Two;
    }
    public bool One { get; }
    public bool Two { get; }
}


public class FooTests
{
    [Theory, AutoData]
    public void GivenOneAndTwo_WhenConstructungFoo(bool one, bool two)
    {
        var subject = new Foo(one, two);

        subject.One.Should().Be(one);
        subject.Two.Should().Be(two);
    }
}
```

What I've ended up doing is implementing the following, which replaces any `AutoData` in the project by default, as I couldn't come up with a reasonable reason I _wouldn't_ want this behaviour:
```C#
public class AutoDataAttribute : AutoFixture.Xunit2.AutoDataAttribute
{
    public AutoDataAttribute() : base(() => new Fixture().Customize(new RandomBooleanSequenceCustomization())){ }
}
```

**Comment by aivascu** [comment, marked as answer]
@airzym AutoFixture is a tool built around the 80-20 rule and explicitly for greenfield TDD.
This means that the default Fixture instance should fit most scenarios and in those cases when it is not sufficient the Fixture can be customized.

It just so happens that alternating boolean values is the most time efficient/performant way to provide pseudo-random values, that fits most scenarios. This is also not the only place where AutoFixture prefers to use non-random value generation. Enum values are also generated in a non-random round-robin fashion.

AutoFixture never claimed to generate permutations for any sets of parameters, any behavior that might have resembled generating permutations was purely accidental.

For scenarios when it is necessary to verify that members are properly initialized, you could use the `ConstructorInitializedMemberAssertion` available in the AutoFixture.Idioms package.
The following test correctly determines that the property `Two` of your SUT is not properly initialized.

```cs
[Theory, AutoData]
public void VerifyCosntructorInitializedMembers(ConstructorInitializedMemberAssertion assertion)
{
    assertion.Verify(typeof(Foo));
}
```

Hope this helps. Let me know if I can help with any other questions.
